### ISSUE TITLE
Query API: Sessions can panic, but can't be caught by middleware

### ISSUE THREAD

**Issue by bobheadxi** [cayleygraph/cayley#807]
**Description**

Queries start a `go session.Execute()` that can't be caught by `defer recover()`-like handlers (since a new goroutine is spawned)

Can be fixed here: https://github.com/cayleygraph/cayley/blob/master/server/http/api_v2.go#L492

**Steps to reproduce the issue:**
1. Run a query that panics
2. Server crashes

**Environment details:**

Backend database: `(database and version)` 0.7.5

Embedded as a library

**Comment by dennwc** [comment]
The panic reason is not relevant in this context, right? (e.g. caused by the application code, not Cayley)

Changing the query API makes sense though. Removing goroutine would help `Execute()`, but now the loop will be in the goroutine, and the same issue will arise. So we should make sequential code possible, similar to how iterators work.

The short-term solution would be to add a panic handler to that goroutine.

**Comment by bobheadxi** [comment]
This particular ticket isn't about the exact issue, which seems to occur in `goja`:

<details><summary>stack</summary>
<code>
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: Panic at 79: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: Panic at 79: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xbe88a7]

goroutine 441 [running]:
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja.(*Runtime).RunProgram.func1(0xc008e0fe58)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja/runtime.go:889 +0x9b
panic(0xe9b5a0, 0xc00b1a4a10)
/usr/local/go/src/runtime/panic.go:522 +0x1b5
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja.(*vm).try.func1(0xc0002cf790, 0x0, 0xc008e0fd58, 0x0, 0x0, 0x0, 0xc008e0fdb0)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja/vm.go:371 +0x48e
panic(0xeb44c0, 0x1bc9220)
/usr/local/go/src/runtime/panic.go:522 +0x1b5
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/cayleygraph/cayley/query/gizmo.(*pathObject).Union(0xc0004486a0, 0x0, 0xc00b80b500)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/cayleygraph/cayley/query/gizmo/traversals.go:295 +0xd7
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/cayleygraph/cayley/query/gizmo.(*pathObject).Or(0xc0004486a0, 0x0, 0x0)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/cayleygraph/cayley/query/gizmo/traversals.go:301 +0x35
reflect.callMethod(0xc00029dd40, 0xc008e0f780, 0xc008e0f768)
/usr/local/go/src/reflect/value.go:690 +0x182
reflect.methodValueCall(0x0, 0x0, 0x0, 0xc00b1a4a00, 0xc008e0f9c0, 0x4bf761, 0xc00b805530, 0xc00029dd40, 0xc00b1a4a00, 0x800000010, ...)
/usr/local/go/src/reflect/asm_amd64.s:35 +0x42
reflect.Value.call(0xe6d8a0, 0xc00029dd40, 0x13, 0xfca913, 0x4, 0xc00b1e7040, 0x1, 0x1, 0xfc7080, 0xc008e0fa01, ...)
/usr/local/go/src/reflect/value.go:447 +0x461
reflect.Value.Call(0xe6d8a0, 0xc00029dd40, 0x13, 0xc00b1e7040, 0x1, 0x1, 0x0, 0x16, 0x0)
/usr/local/go/src/reflect/value.go:308 +0xa4
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja.(*Runtime).wrapReflectFunc.func1(0x12a0600, 0xc0004486e0, 0xc00031afe0, 0x1, 0xc, 0x12a0ba0, 0x1c00dc0)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja/runtime.go:1171 +0x791
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja.(*vm)._nativeCall(0xc0002cf790, 0xc000445760, 0x1)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja/vm.go:1833 +0x29a
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja.call.exec(0x1, 0xc0002cf790)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja/vm.go:1817 +0x4a4
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja.(*vm).run(0xc0002cf790)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja/vm.go:290 +0x95
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja.(*vm).try(0xc0002cf790, 0xc008e0fdb8, 0x0)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja/vm.go:377 +0x10d
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja.(*vm).runTry(0xc0002cf790, 0x1009820)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja/vm.go:382 +0x4e
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja.(*Runtime).RunProgram(0xc00a5ec600, 0xc000585080, 0x0, 0x0, 0x0, 0x0)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/dop251/goja/runtime.go:900 +0x1c9
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/cayleygraph/cayley/query/gizmo.(*Session).run(0xc008ca0300, 0xc000340480, 0x20f, 0xc000584ba0, 0xc008ca0300, 0xc0001330e0, 0xc00060d730)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/cayleygraph/cayley/query/gizmo/gizmo.go:243 +0x6a
go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/cayleygraph/cayley/query/gizmo.(*Session).Execute(0xc008ca0300, 0x128e020, 0xc000584ba0, 0xc000340480, 0x20f, 0xc000585020, 0x64)
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/cayleygraph/cayley/query/gizmo/gizmo.go:266 +0x172
created by go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/cayleygraph/cayley/server/http.(*APIv2).ServeQuery
/go/src/go.riotgames.com/rpclin/alerts-explorer/vendor/github.com/cayleygraph/cayley/server/http/api_v2.go:457 +0x3d3
</code>
</details>

But yeah I would like to add a short-term handler to the goroutine, like:

```go
go func() {
    defer ... // handler
    session.Execute()
}()
```

**Comment by dennwc** [comment]
> I would like to add a short-term handler to the goroutine

I'm okay with accepting this change, feel free to send a PR if you have some spare time :)

It will be necessary to pass the error back from the goroutine, so the panic is not discarded silently. Maybe adding it to the [implementation](https://github.com/cayleygraph/cayley/blob/master/query/gizmo/gizmo.go#L251) will be a good idea? In this case, we can store an error to the `Session` and return it in `Result.Err()`, for example.
As a part of this short-term solution, we can also document that `Execute` must handle panics internally.

**Comment by bobheadxi** [comment]
Actually, I don't think a fix here is needed - it looks like the panic occured in `gizmo.(*Session).run`, which doesn't spawn a goroutine that might cause this issue. the `goja` entrypoint at `goja.(*Runtime).RunProgram` [already has a recover](https://sourcegraph.com/github.com/dop251/goja/-/blob/runtime.go#L884)

The bug itself seems to be caused during union on a nil pointer:


```
github.com/cayleygraph/cayley/query/gizmo.(*pathObject).Union(0xc0004486a0, 0x0, 0xc00b80b500)
github.com/cayleygraph/cayley/query/gizmo/traversals.go:295 +0xd7
github.com/cayleygraph/cayley/query/gizmo.(*pathObject).Or(0xc0004486a0, 0x0, 0x0)
github.com/cayleygraph/cayley/query/gizmo/traversals.go:301 +0x35
```

which seems to have been fixed in https://github.com/cayleygraph/cayley/commit/c78fbee8a0958d59b1a8f36ff78c0b97306fabd9#diff-d6d007fe34426cf4bae98518c94d7865R284

are there plans for a release soon? I have a feeling this issue is already resolved

**Comment by dennwc** [comment]
> are there plans for a release soon?

Yes, but no ETA at the moment. I suggest building from that commit for now.

**Comment by dennwc** [comment]
Closing since it's fixed in `master` and will be released as a part of v0.7.6.

Also note that we [changed the query interface](https://github.com/cayleygraph/cayley/commit/1884ec7e960cd5e75bd7e9c0bf04bd5c840c7181) as well recently, in case you run Cayley from `master`.
